apiserver/Makefile

```diff
@@ -43,16 +17,86 @@ jupyter_notebook_config.py:
 	    -e "s,@keyfile@,$(GSA_KEY_FILE),g" \
             < jupyter_notebook_config.py.in > $@
 
-$(eval $(call DOCKER_IMAGE,spark-base,install-hail-locally jupyter_notebook_config.py,SPARK_BASE_IMAGE))
-$(eval $(call DOCKER_IMAGE,hail-base,install-hail-locally jupyter_notebook_config.py,HAIL_BASE_IMAGE))
-$(eval $(call DOCKER_IMAGE,spark-master,install-hail-locally jupyter_notebook_config.py,SPARK_MASTER_IMAGE))
-$(eval $(call DOCKER_IMAGE,spark-worker,install-hail-locally jupyter_notebook_config.py,SPARK_WORKER_IMAGE))
-$(eval $(call DOCKER_IMAGE,apiserver,install-hail-locally jupyter_notebook_config.py,APISERVER_IMAGE))
-$(eval $(call DOCKER_IMAGE,hail-jupyter,install-hail-locally jupyter_notebook_config.py,HAIL_JUPYTER_IMAGE))
+ifeq ($(IN_HAIL_CI),1)
+.PHONY: build-spark-base build-hail-base build-spark-master build-spark-worker build-apiserver
+
+build-spark-base:
+	docker pull debian:9.5
+	-docker pull $(REPO)/spark-base:latest
+	docker build -t spark-base -f Dockerfile.spark-base . --cache-from spark-base,$(REPO)/spark-base:latest,debian:9.5
+
+build-hail-base: build-spark-base install-hail-locally
+	-docker pull $(REPO)/spark-base:latest
+	-docker pull $(REPO)/hail-base:latest
+	docker build -t hail-base -f Dockerfile.hail-base . --cache-from hail-base,$(REPO)/hail-base:latest,spark-base
+
+build-spark-master: build-hail-base
+	-docker pull $(REPO)/hail-base:latest
+	-docker pull $(REPO)/spark-master:latest
+	docker build -t spark-master -f Dockerfile.spark-master . --cache-from spark-master,$(REPO)/spark-master:latest,hail-base
+
+build-spark-worker: build-hail-base
+	-docker pull $(REPO)/hail-base:latest
+	-docker pull $(REPO)/spark-worker:latest
+	docker build -t spark-worker -f Dockerfile.spark-worker . --cache-from spark-worker,$(REPO)/spark-worker:latest,hail-base
+
+build-apiserver: build-hail-base
+	-docker pull $(REPO)/hail-base:latest
+	-docker pull $(REPO)/apiserver:latest
+	docker build -t apiserver -f Dockerfile.apiserver . --cache-from apiserver,$(REPO)/apiserver:latest,hail-base
+
+build-hail-jupyter: build-hail-base jupyter_notebook_config.py
+	-docker pull $(REPO)/hail-base:latest
+	-docker pull $(REPO)/hail-jupyter:latest
+	docker build -t hail-jupyter -f Dockerfile.hail-jupyter . --cache-from hail-jupyter,$(REPO)/hail-jupyter:latest,hail-base
+
+build: build-spark-master build-spark-worker build-apiserver build-hail-jupyter
+
+SPARK_MASTER_IMAGE = $(REPO)/spark-master:$(shell docker images -q --no-trunc spark-master | sed -e 's,[^:]*:,,')
+SPARK_WORKER_IMAGE = $(REPO)/spark-worker:$(shell docker images -q --no-trunc spark-worker | sed -e 's,[^:]*:,,')
+APISERVER_IMAGE = $(REPO)/apiserver:$(shell docker images -q --no-trunc apiserver | sed -e 's,[^:]*:,,')
+HAIL_JUPYTER_IMAGE = $(REPO)/hail-jupyter:$(shell docker images -q --no-trunc hail-jupyter | sed -e 's,[^:]*:,,')
+
+push: build
+	docker tag spark-master $(SPARK_MASTER_IMAGE)
+	docker push $(SPARK_MASTER_IMAGE)
+	docker tag spark-master $(REPO)/spark-master
```

Comment:
We're putting :latest explicitly on repositories (like the pull code above)

Reply:
done